### PR TITLE
RResultPtr<T>::Release method added to RResultPtr.hxx

### PR DESCRIPTION
**Pull request by Harshalzzzzzzz**
Dear Mentor,
Here's my pull request for the issue https://github.com/root-project/root/issues/7047.
Can you please review this. I will squash the commits after the review.

Thanks in advance !
@eguiraud

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsft-bot build!

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 1]
```diff
@@ -170,7 +170,15 @@ public:
 
    /// Get a const reference to the encapsulated object.
    /// Triggers event loop and execution of all actions booked in the associated RLoopManager.
-   const T &GetValue() { return *Get(); }
+   const T &GetValue()
+   {
+       try{
+           if( fObjPtr != nullptr)
+               throw fObjPtr;
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RResultPtr.hxx:177]: in C++ you don't throw objects, this should throw a `std::runtime_error`. Also we shouldn't try-catch our own throw, and the error condition is `fObjPtr == nullptr`

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 2]
```diff
@@ -302,6 +310,16 @@ public:
       return *this;
    }
 
+    /// Release method for RResultPtr.hxx, similar to std::shared_ptr to release ownership
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RResultPtr.hxx:313]:
```suggestion
    /// Return a pointer to the result, releasing its ownership and leaving this object empy. 
```

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 3]
```diff
@@ -302,6 +310,16 @@ public:
...
+    RResultPtr<T> &Release()
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RResultPtr.hxx:314]: This should return a `T*`, not a `RResultPtr<T>`.

File: tree/dataframe/test/dataframe_resptr.cxx [hunk 4]
```diff
@@ -72,6 +72,15 @@ TEST(RResultPtr, ImplConv)
    EXPECT_TRUE(hasRun);
 }
 
+TEST(RResultPtr, Release)
+{
+    ROOT::RDataFrame df(1);
+    auto p = df.Sum<ULong64_t>("rdfentry_");
+    p.GetValue();
+    EXPECT_TRUE(p.Release());
+    EXPECT_TRUE(p != nullptr);
```
Comment (eguiraud) [inline tree/dataframe/test/dataframe_resptr.cxx:81]:
```suggestion
    EXPECT_TRUE(p == nullptr);
```

File: tree/dataframe/test/dataframe_resptr.cxx [hunk 5]
```diff
@@ -72,6 +72,15 @@ TEST(RResultPtr, ImplConv)
    EXPECT_TRUE(hasRun);
 }
 
+TEST(RResultPtr, Release)
+{
+    ROOT::RDataFrame df(1);
+    auto p = df.Sum<ULong64_t>("rdfentry_");
+    p.GetValue();
+    EXPECT_TRUE(p.Release());
```
Comment (eguiraud) [inline tree/dataframe/test/dataframe_resptr.cxx:80]: what is this checking?

**Review by eguiraud — CHANGES_REQUESTED** [review]
Hi @Harshalzzzzzzz ,
thank you for the PR, please fix the compilation errors and address my inline comments.

Cheers,
Enrico

File: tree/dataframe/inc/ROOT/RResultPtr.hxx (hunk 1 above)
Comment (Harshalzzzzzzz) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx:177]: Since you had mentioned about throwing exception, can you suggest what try catch can be replaced with.

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx:177]: Sure, not problem!

Library code just throws the error. Client code can catch if it wants. I.e. there should be no try/catch here, just the throw

Comment (Harshalzzzzzzz) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx]: @eguiraud Review Changes Completed
Thank you !

**Comment by bellenot** [comment]
@phsft-bot build

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 6]
```diff
@@ -302,6 +307,16 @@ public:
       return *this;
    }
 
+    /// Return a pointer to the result, releasing its ownership and leaving this object empy. 
+    T* &Release()
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RResultPtr.hxx:311]: Why a reference to pointer?

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 7]
```diff
@@ -302,6 +307,16 @@ public:
...
+    {
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RResultPtr.hxx:312]: We should make sure to trigger the generation of the result first:

```suggestion
   {
      if (!fActionPtr->HasRun())
         TriggerRun(); 
```

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 8]
```diff
@@ -170,7 +170,12 @@ public:
 
    /// Get a const reference to the encapsulated object.
    /// Triggers event loop and execution of all actions booked in the associated RLoopManager.
-   const T &GetValue() { return *Get(); }
+   const T &GetValue()
+   {
+       if( fObjPtr == nullptr)
+           throw;
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RResultPtr.hxx:176]:
```suggestion
           throw std::runtime_error("Trying to access the contents of a null RResultPtr.");
```

**Review by eguiraud — CHANGES_REQUESTED** [review]
Hi @Harshalzzzzzzz , alright, almost there :smile: 

Please use the standard ROOT indentation of 3 spaces (I know, weird convention). The easiest way is to run `clang-format` over your changes.

Also see some inline comments.

**Comment by Harshalzzzzzzz** [comment]
@eguiraud Sure, thank you for the inputs. 
Will stick to the expected standards and conventions for future contributions.

File: tree/dataframe/inc/ROOT/RResultPtr.hxx (hunk 6 above)
Comment (Harshalzzzzzzz) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx:311]: I added it so, while keeping in mind the way other methods have been implemented for RResultPtr, just in terms of standards

File: tree/dataframe/inc/ROOT/RResultPtr.hxx (hunk 8 above)
Comment (Harshalzzzzzzz) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx]: Yes this seems better, I thought of only terminating it for that condition so I added only `throw ;`

**Comment by eguiraud** [comment]
@phsft-bot build

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 9]
```diff
@@ -302,6 +307,18 @@ public:
       return *this;
    }
 
+    /// Return a pointer to the result, releasing its ownership and leaving this object empty.
+    T* &Release()
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RResultPtr.hxx:311]:
```suggestion
    T* Release()
```

Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx:311]: this removes the compilation warnings

Comment (Harshalzzzzzzz) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx]: @eguiraud  Previous Version of the commit wasn't pushed
I have fixed it now.

**Comment by eguiraud** [comment]
@phsft-bot build

**Review by eguiraud — CHANGES_REQUESTED** [review]
@Harshalzzzzzzz alright, the changes look good to me, there are just some formatting issues (e.g. whitespaces in wrong places). Please run `git clang-format master` and add a commit with the formatting fixes.

**Comment by eguiraud** [comment]
@Harshalzzzzzzz sorry you might have given the wrong command or similar because this PR now touches 229 files (!).

`git clang-format master` should just fix the formatting of the diff between your branch and master. If that's not what happens, you can run `clang-format` manually on `RResultPtr.hxx` and commit the changes that fix the formatting of your patch.

**Comment by Harshalzzzzzzz** [comment]
@eguiraud Sorry about the issue, I had selected wrong rebase version while trying to squash commits, my bad. PR is Fixed now

**Comment by eguiraud** [comment]
@phsft-bot build

**Review by eguiraud — APPROVED** [review]
The Windows build failure is unrelated. This is good to merge, thanks @Harshalzzzzzzz !

**Pull request merged**